### PR TITLE
Hackathon: Post Reminders

### PR DESCRIPTION
**Pull request by ashishbhate**
#### Summary

- Adds the frontend UI for post reminders. Most of the actual UI is copied from the DND status feature. The components in the DND status feature weren't re-usable. 

#### Ticket Link

- N/A

#### Related Pull Requests

- Has server changes - https://github.com/mattermost/mattermost-server/pull/20555

#### Release Note

```release-note
Added the frontend changes for the post reminders feature
```

### CODE REVIEW

File: components/dot_menu/dot_menu.tsx
Comment (ashishbhate) [inline components/dot_menu/dot_menu.tsx:632]: When the window size is narrow, the submenu get hidden behind the main channel view when opened in the RHS. 

This is because of https://github.com/mattermost/dynamic-virtualized-list/blob/master/src/DynamicSizeList.js#L371

If we remove that line, then the RHS won't scroll.

The current fix is for the menu to open on top of the dot menu when opened in the RHS.

File: components/widgets/menu/menu_items/submenu_item.tsx
Comment (ashishbhate) [inline components/widgets/menu/menu_items/submenu_item.tsx:140]: There was a prop called `subMenuClass` but it was unused. Using it here. Not sure why it was unused, this is one of the reasons the CI tests are failing.

File: components/widgets/menu/menu_items/submenu_item.tsx
Comment (ashishbhate) [inline components/widgets/menu/menu_items/submenu_item.tsx:178]: From the docs for [classNames](https://www.npmjs.com/package/classnames), the `{styleSelectableItem}` doesn't do anything. I haven't changed it. Just highlighting what seems to be dead code.

File: components/dot_menu/dot_menu.tsx
Comment (ashishbhate) [reply components/dot_menu/dot_menu.tsx:632]: **In center channel:**

Screenshot 2022-07-05 at 3 50 27 PM

(this also hides behind the LHS, if the center channel view is small enough)

**RHS hides behind center channel**
Screenshot 2022-07-05 at 3 53 37 PM


**With the fix, it overlaps the dot menu**
Screenshot 2022-07-05 at 3 50 48 PM

Comment (agnivade) [reply components/dot_menu/dot_menu.tsx:632]: Possibly a dumb suggestion from a backend person: can we apply some `!important` hackery to override the library's CSS?

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
This is awesome to see this getting in to the product! Great work. 

A few things from my viewpoint:

1. Can we change the title in the submenu from "Remind after:" to "Set a reminder for:"
2. In the Custom time modal, change the title to "Set a reminder". 
3. In the Custom time modal, it looks like we can tighten up some of the space under the title - should match what we have for the custom DND time modal
4. In the Custom time modal, make sure the time format is shown based on the time display setting for the user. Right now, it appears to always show a 24-hour clock format (we should fix this in the DND modal too, but can be another ticket)
5. For the message that is posted when the reminder is set, as you mentioned, we should remove the line breaks. Maybe we should consider adjusting the message to "You will be reminded at {time}, {date} about this message from {@username/display name}: {permalink}"
6. The permalink preview that appears in the ephemeral message that confirms the reminder is set, has its text color greyed out. I assume it's getting inherited from the standard system message body text color or something. Can we change that so the permalink preview doesn't get greyed out but only the actual system message does?
7. Also, as you mentioned, the submenu in the RHS doesn't feel right as it overlaps most of the menu under it rather than displaying to the left of it
8. For the message that is posted by the systembot when the reminder executes, can we update the message to "Hi there, here's your reminder about this message from {@username/display name}: {permalink}"
9. This is more of a problem with the behaviors for submenus, but I find it hard to get to the submenu sometimes - especially if I'm trying to right to one of the bottom options in the submenu. The submenu closes when I diagonally move my mouse from the parent menu to the child menu. Again, this is more of a problem with parent/child menus we have. We would need to look at a holistic solution to this rather than just for this specific case.....I wanted to call it out though since I struggled with choosing the last few options in the submenu a couple times.

**To consider as enhancements:**

10. Is there some way to view all my active reminders? If we can consider some way to view all active reminders in an RHS or something, that would be cool. Then you could edit/delete/manage any of your reminders as well.
11. We could consider adding action buttons to the confirmation message when a reminder is set to [Undo] or [View all my reminders]
12. If you create a reminder for the same post, should we consider warning the user that they have an active reminder for the same post already and confirm if they wish to creat another?
13. When the systembot sends the reminder, should we consider adding a snooze button with options to snooze for 5 minutes, 30 minutes, 1 hr, 2 hrs, tomorrow at same time

File: components/dot_menu/dot_menu.tsx
Comment (ashishbhate) [reply components/dot_menu/dot_menu.tsx:632]: As far as I understand it, `!important` lets a parent overwrite a child's CSS. Here we want to the child to overwrite the parent, but that doesn't seem to be working, and forcing it in the parent breaks scrolling. 

I tried setting `!important` but I couldn't get it to work.

File: components/widgets/menu/menu_items/submenu_item.tsx
Comment (ashishbhate) [reply components/widgets/menu/menu_items/submenu_item.tsx:178]: Yep, but since `styleSelectableItem` is a boolean, it just adds the class `.styleSelectableItem` which isn't used anywhere. 

The prop is correctly used in this line https://github.com/mattermost/mattermost-webapp/blob/master/components/widgets/menu/menu_items/submenu_item.tsx#L149

The way it's used here doesn't do anything.

**Review by rmonga71 — COMMENTED** [review]
+1 to @matthewbirtch 's feedback and proposed enhancements. Few comments:
- I noticed that pressing "x" on the system message that is generated when you create a reminder hides the message (as expected) but doesn't cancel the reminder - how do we feel about changing that behaviour? Once the system message is closed, there's nothing to indicate that you have a reminder for a given message. 
- Thinking out loud, some other potential solutions: visual indicator like a badge on a message with a reminder (I know there's lots of badge work pending with MPA etc. so I understand not wanting to overcrowd), or maybe the system-bot sends a confirmation message ("Hey there, I'll remind you here about message "abcd" on `date` at `time`").

3/5, I do feel that some method to cancel reminders or at least indicate which messages have reminders is necessary before merging.

**Comment by agnivade** [comment]
@rmonga71 : IMO, I think cancelling reminders can be done as part of a later PR. However, I like your second option to have the bot post a DM message about the reminder rather than an ephemeral message in channel. This way the message persists, and the user can easily see what messages have reminders set.

@matthewbirtch : 

> Is there some way to view all my active reminders? If we can consider some way to view all active reminders in an RHS or something, that would be cool. Then you could edit/delete/manage any of your reminders as well.

There isn't, as of now, but can be easily added later. A simple solution is to have a dedicated slash command with `/reminders list`, but wondering if we should have a dedicated UI section like - threads, insights?

> We could consider adding action buttons to the confirmation message when a reminder is set to [Undo] or [View all my reminders]

Definitely. This goes well with @rmonga71's second suggestion to change from an in-channel ephemeral message to a bot DM, in the same channel where the reminders are actually sent.

> If you create a reminder for the same post, should we consider warning the user that they have an active reminder for the same post already and confirm if they wish to creat another?

We discussed this, and the behavior we went with was to extend the reminder of the current post to the new time.

> When the systembot sends the reminder, should we consider adding a snooze button with options to snooze for 5 minutes, 30 minutes, 1 hr, 2 hrs, tomorrow at same time

Yep, defintely a great enhancement that can be done.

File: components/dot_menu/dot_menu.tsx
Comment (ashishbhate) [reply components/dot_menu/dot_menu.tsx]: Done

File: components/post_reminder_time_picker_modal/post_reminder_custom_time_picker_modal.tsx
Comment (ashishbhate) [reply components/post_reminder_time_picker_modal/post_reminder_custom_time_picker_modal.tsx:40]: I don't mind making the change, but I'm not sure I understand why it would be good to do so? Here we only need the current date to set in the list.

File: components/post_reminder_time_picker_modal/post_reminder_custom_time_picker_modal.tsx
Comment (ashishbhate) [reply components/post_reminder_time_picker_modal/post_reminder_custom_time_picker_modal.tsx:44]: This is just a small helper, if the time is past 11:30 PM then we set the date to tomorrow and the time to 00:00 (midnight).

**Comment by ashishbhate** [comment]
@matthewbirtch 

> 1. Can we change the title in the submenu from "Remind after:" to "Set a reminder for:"

Done
> 2. In the Custom time modal, change the title to "Set a reminder".

Done

> 3. In the Custom time modal, it looks like we can tighten up some of the space under the title - should match what we have for the custom DND time modal

Done

> 4. In the Custom time modal, make sure the time format is shown based on the time display setting for the user. Right now, it appears to always show a 24-hour clock format (we should fix this in the DND modal too, but can be another ticket)

Done, also fixed in the ephemeral post when the reminder is set

> 5. For the message that is posted when the reminder is set, as you mentioned, we should remove the line breaks. Maybe we should consider adjusting the message to "You will be reminded at {time}, {date} about this message from {@username/display name}: {permalink}"

Done

> 6. The permalink preview that appears in the ephemeral message that confirms the reminder is set, has its text color greyed out. I assume it's getting inherited from the standard system message body text color or something. Can we change that so the permalink preview doesn't get greyed out but only the actual system message does?

Done

> 8. For the message that is posted by the systembot when the reminder executes, can we update the message to "Hi there, here's your reminder about this message from {@username/display name}: {permalink}"

Done in server PR https://github.com/mattermost/mattermost-server/pull/20733

> 7. Also, as you mentioned, the submenu in the RHS doesn't feel right as it overlaps most of the menu under it rather than displaying to the left of it

> 9. This is more of a problem with the behaviors for submenus, but I find it hard to get to the submenu sometimes - especially if I'm trying to right to one of the bottom options in the submenu. The submenu closes when I diagonally move my mouse from the parent menu to the child menu. Again, this is more of a problem with parent/child menus we have. We would need to look at a holistic solution to this rather than just for this specific case.....I wanted to call it out though since I struggled with choosing the last few options in the submenu a couple times.

Both of these need changes to some of the fundamental building blocks of the webapp. I haven't been able to do them here yet. I'm going to chat with @koox00 and @AshishDhama to figure out the best way to make it work.

**Comment by ashishbhate** [comment]
@matthewbirtch @rmonga71 @agnivade

Thanks for the feedback and ideas for improvements. As Agniva already suggested, I'd like to get this PR in first, and then work on feature additions incrementally. This PR is already quite big and changes a lot of existing code. 

Also, the implementation in this PR, the suggestions in this PR, and the preliminary feature design https://mattermost.atlassian.net/browse/MM-36130 don't exactly align. 

After this PR is done, I'd like to reconcile the different ideas and break that jira ticket down into actionable tasks for devs. 

// cc @esethna

**Comment by ashishbhate** [comment]
@koox00 @AshishDhama This is ready for another round of dev reviews. It would be great if you guys could help figuring the menu overlap issue (I've started a discussion above) and the sub-menu hover issue that Matt mentioned.

**Comment by agnivade** [comment]
Going to add load-test coverage today. Please do not merge before that :)

File: components/widgets/menu/menu_items/submenu_item.tsx
Comment (koox00) [reply components/widgets/menu/menu_items/submenu_item.tsx:178]: I see a class named `styleSelectableItem` in `menu.scss` https://github.com/mattermost/mattermost-webapp/blob/master/components/widgets/menu/menu.scss#L107,
isn't it applied and used here?

File: components/post_reminder_time_picker_modal/post_reminder_custom_time_picker_modal.tsx
Comment (koox00) [reply components/post_reminder_time_picker_modal/post_reminder_custom_time_picker_modal.tsx:44]: I don't see where we set the time to 00:00

File: components/post_reminder_time_picker_modal/post_reminder_custom_time_picker_modal.tsx
Comment (koox00) [reply components/post_reminder_time_picker_modal/post_reminder_custom_time_picker_modal.tsx:40]: I remember that the time picker was off for my timezone, let me check again

**Comment by esethna** [comment]
@ashishbhate is there a feature flag? Given it hasn't "passed" UX/PM review before this will get merged should we add a feature flag to ensure we can review an iterate before it ships?

**Comment by agnivade** [comment]
We decided not to add a feature flag since this feature is anyways an opt-in. But given that there is work to be done from a design perspective, it makes sense to add one and start iterating on the feature.

File: components/post_reminder_time_picker_modal/post_reminder_custom_time_picker_modal.tsx
Comment (ashishbhate) [reply components/post_reminder_time_picker_modal/post_reminder_custom_time_picker_modal.tsx:44]: >I don't see where we set the time to 00:00

Its on line 109 to line 124

File: components/widgets/menu/menu_items/submenu_item.tsx
Comment (ashishbhate) [reply components/widgets/menu/menu_items/submenu_item.tsx:178]: Ah crap! I didn't think to look in the css files. That clears it up, thanks!

**Comment by ashishbhate** [comment]
>@ashishbhate is there a feature flag? Given it hasn't "passed" UX/PM review before this will get merged should we add a feature flag to ensure we can review and iterate before it ships?

Yep, makes sense. I'll add a feature flag.

**Comment by matthewbirtch** [comment]
> After this PR is done, I'd like to reconcile the different ideas and break that jira ticket down into actionable tasks for devs.

Works for me @ashishbhate. Keep me posted on this effort as I'd be happy to help spec out the ideas.

**Comment by ashishbhate** [comment]
@AshishDhama gentle bump for your review and thoughts on the submenu overlap

**Comment by AshishDhama** [comment]
@ashishbhate on clicking remind it’s opening RHS

https://user-images.githubusercontent.com/16203333/189613145-bda6ab8c-033b-46b7-bb14-df156f5d3ae1.mov

**Comment by AshishDhama** [comment]
@ashishbhate open sub-menu on hover is great, but I think open on click, will resolve the overlap issue cc// @matthewbirtch 
Maybe we can dismiss the submenu by clicking outside(a grey area) and take the user back to the main menu.
Screenshot 2022-09-12 at 2 26 16 PM

**Review by AshishDhama — COMMENTED** [review]
The rest of the code is looking good to me so far but will through it again tonight.

**Comment by matthewbirtch** [comment]
> @ashishbhate opening sub-menu on hover is great, but I think opening it on click will resolve the overlap issue cc// @matthewbirtch Maybe we can dismiss the submenu by clicking outside(a grey area) and take the user back to the main menu. Screenshot 2022-09-12 at 2 26 16 PM

This feels like a band-aid solution to me personally. We are show submenus elsewhere in the product on hover (e.g. in the LHS channel context menu, or in the Channel header dropdown menu) so I wouldn't want to change this behavior. I don't fully understand the reasons why we can't make it work as intended though, so forgive me.

**Comment by esethna** [comment]
@matthewbirtch we discussed this at Mattercon. Essentially the way the app layers panes (Global header > sidebar > center channel > RHS) causes the sub menu in the RHS to always be below the center channel. We figured the best approach for now is to push the submenu as far left as possible without it going under the center channel. It's not ideal (since the sub menu might still overlap the ... menu especially on smaller screen widths, but it allows us to ship this before the web platform team properly fixes the menu component with an overhaul in Q4. Example:
image

cc// @ashishbhate

**Comment by matthewbirtch** [comment]
> We figured the best approach for now is to push the submenu as far left as possible without it going under the center channel. It's not ideal (since the sub menu might still overlap the ... menu especially on smaller screen widths, but it allows us to ship this before the web platform team properly fixes the menu component with an overhaul in Q4. 

Ok, that makes sense. I agree it's not ideal, but as long as all the items in the submenu are still accessible and we're not obscuring any labels from the main menu, we can probably live with this until it's properly fixed.

I wonder if we should also consider reducing the width of the parent menu slightly. It looks like we might be able to reduce the width by 20-30 pixels. That might help a bit

**Comment by AshishDhama** [comment]
closing this in favour of https://github.com/mattermost/mattermost-webapp/pull/12007

**Pull request closed without merging**
